pkg/utils/cipher.go

```diff
@@ -55,12 +55,100 @@ func CryptoConfigWrapper(encBlock *evecommon.EncryptionBlock, cmnCryptoCfg *Comm
 
 	mEncBlock, mErr := proto.Marshal(encBlock)
 	if mErr != nil {
-		return nil, fmt.Errorf("error marshalling user data: %v", mErr)
+		return nil, fmt.Errorf("error marshalling user data: %w", mErr)
 	}
-	//Fill CipherBlock.
+	// Fill CipherBlock.
 	cipherBlock, cbErr := createCipherBlock(mEncBlock, cipherCtx.ContextId, cmnCryptoCfg, iv[:16])
 	if cbErr != nil {
-		return nil, fmt.Errorf("error creating cipher block: %v", cbErr)
+		return nil, fmt.Errorf("error creating cipher block: %w", cbErr)
 	}
 	return cipherBlock, nil
 }
+
+// internal decryption method
+func aesDecrypt(iv, symmetricKey, ciphertext []byte) ([]byte, error) {
+	plaintext := make([]byte, len(ciphertext))
+	aesBlockDecrypter, err := aes.NewCipher(symmetricKey)
+	if err != nil {
+		return nil, err
+	}
+	aesDecrypter := cipher.NewCFBDecrypter(aesBlockDecrypter, iv)
+	aesDecrypter.XORKeyStream(plaintext, ciphertext)
+	return plaintext, nil
+}
+
+// reverse the process of creating a cipher block
+func decryptCipherBlock(cipherBlock *evecommon.CipherBlock, cmnCryptoCfg *CommonCryptoConfig) ([]byte, error) {
+	if cipherBlock == nil {
+		return nil, fmt.Errorf("nil cipher block in decrypt cipher block method")
+	}
+
+	// decrypt ciphertext using symmetric key and initial value
+	decryptedText, decErr := aesDecrypt(cipherBlock.InitialValue, cmnCryptoCfg.SymmetricKey, cipherBlock.CipherData)
+	if decErr != nil {
+		return nil, decErr
+	}
+
+	// verify sha256 checksum of decrypted text
+	shaOfDecryptedText := sha256.Sum256(decryptedText)
+	if !sha256Equal(shaOfDecryptedText[:], cipherBlock.ClearTextSha256) {
+		return nil, fmt.Errorf("decrypted text does not match original sha256 checksum")
+	}
+
+	return decryptedText, nil
+}
+
+// helper function to compare two sha256 hashes
+func sha256Equal(a, b []byte) bool {
```

Comment:
how about making this generic and putting this to `pkg/utils/slices.go`, i.e.

```golang

func CompareSlices[T comparable](lhs, rhs []T) bool {
  if len(lhs) != len(rhs) {
    return false
  }

  for i := range lhs {
    if lhs[i] != rhs[i] {
       return false
  }

   return true
}

```

Reply:
Thanks, that's much better!